go/vt/sqlparser/ast.go

```diff
@@ -1746,12 +1746,17 @@ type ColumnType struct {
 	Scale    *Literal
 
 	// Text field options
-	Charset string
+	Charset ColumnCharset
 
 	// Enum values
 	EnumValues []string
 }
 
+type ColumnCharset struct {
+	Name   string
+	Binary bool
+}
+
```

Comment:
This type was added because in the type definition it's possible to add the binary marker for a character set, so we need to track this. We can't at the point of where we parse things backfill this into the existing collation. Firstly because we don't have access to that node, but more importantly because it would generate syntax that is invalid.

So we must track this as a separate property.

See also https://dev.mysql.com/doc/refman/8.0/en/cast-functions.html and specifically:

```
A COLLATE clause is not permitted within a [CONVERT()](https://dev.mysql.com/doc/refman/8.0/en/cast-functions.html#function_convert) or [CAST()](https://dev.mysql.com/doc/refman/8.0/en/cast-functions.html#function_cast) call, but you can apply it to the function result. For example, these are legal:
```

But syntax like the following is valid:

```
SELECT CONVERT('test', CHAR CHARACTER SET utf8mb4 binary)
```

Which will end up using the `utf8mb4_bin` collation (see also the evalengine changes in this PR).

Reply:
nit: let's merge this block with the block above

Reply:
> This type was added because in the type definition it's possible to add the binary marker for a character set, so we need to track this. We can't at the point of where we parse things backfill this into the existing collation. Firstly because we don't have access to that node, but more importantly because it would generate syntax that is invalid.

Makes sense, thank you!

Reply:
I think we should write a code comment here that contains what you wrote in your GitHub comment:
> This type was added because in the type definition it's possible to add the binary marker for a character set, so we need to track this. We can't at the point of where we parse things backfill this into the existing collation. Firstly because we don't have access to that node, but more importantly because it would generate syntax that is invalid.

Reply:
> nit: let's merge this block with the block above

Not entirely sure what you mean by this?

Reply:
Rest is addressed!